File: bkmonitor/webpack/src/rum/src/plugins/web-vitals.ts
/*
 * Tencent is pleased to support the open source community by making
 * 蓝鲸智云PaaS平台 (BlueKing PaaS) available.
 *
 * Copyright (C) 2017-2025 Tencent.  All rights reserved.
 *
 * 蓝鲸智云PaaS平台 (BlueKing PaaS) is licensed under the MIT License.
 *
 * License for 蓝鲸智云PaaS平台 (BlueKing PaaS):
 *
 * ---------------------------------------------------
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation
 * the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and
 * to permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of
 * the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
 * THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF
 * CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

import type { BkOTRumConfig } from '../core/config';
import type { BkOTPlugin } from '../core/plugin';
import type { Attributes } from '@opentelemetry/api';

interface VitalLike {
  attribution?: Record<string, unknown>;
  id?: string;
  name: string;
  rating?: string;
  value: number;
}

const getNavigationType = () => {
  const navigation = performance.getEntriesByType('navigation')[0] as PerformanceNavigationTiming | undefined;

  return navigation?.type || 'unknown';
};

// 仅挑选低基数、可枚举的字段进 metric attributes，避免指标维度爆炸
const getMetricDimensions = (metric: VitalLike): Attributes => ({
  'rum.navigation.type': getNavigationType(),
  'web_vital.name': metric.name,
  'web_vital.rating': metric.rating,
});

// 把 attribution 中价值最高的字段抽出来（仅放进 span/log，不放进 metric）
const pickAttribution = (metric: VitalLike, redactUrl: (url: string) => string): Attributes => {
  const attr = metric.attribution ?? {};
  const result: Attributes = {};

  const setIfDefined = (key: string, value: unknown, transform?: (value: string) => string) => {
    if (value === undefined || value === null) return;
    if (typeof value === 'string' || typeof value === 'number' || typeof value === 'boolean') {
      result[key] = typeof value === 'string' && transform ? transform(value) : value;
    }
  };

  switch (metric.name) {
    case 'LCP':
      setIfDefined('web_vital.lcp.url', attr.url, redactUrl);
      setIfDefined('web_vital.lcp.target', attr.target);
      setIfDefined('web_vital.lcp.element_render_delay', attr.elementRenderDelay);
      setIfDefined('web_vital.lcp.resource_load_duration', attr.resourceLoadDuration);
      setIfDefined('web_vital.lcp.time_to_first_byte', attr.timeToFirstByte);
      break;
    case 'CLS':
      setIfDefined('web_vital.cls.largest_shift_target', attr.largestShiftTarget);
      setIfDefined('web_vital.cls.largest_shift_value', attr.largestShiftValue);
      setIfDefined('web_vital.cls.load_state', attr.loadState);
      break;
    case 'INP':
      setIfDefined('web_vital.inp.interaction_target', attr.interactionTarget);
      setIfDefined('web_vital.inp.interaction_type', attr.interactionType);
      setIfDefined('web_vital.inp.input_delay', attr.inputDelay);
      setIfDefined('web_vital.inp.processing_duration', attr.processingDuration);
      setIfDefined('web_vital.inp.presentation_delay', attr.presentationDelay);
      break;
    case 'FCP':
      setIfDefined('web_vital.fcp.time_to_first_byte', attr.timeToFirstByte);
      setIfDefined('web_vital.fcp.load_state', attr.loadState);
      break;
    case 'TTFB':
      setIfDefined('web_vital.ttfb.waiting_duration', attr.waitingDuration);
      setIfDefined('web_vital.ttfb.dns_duration', attr.dnsDuration);
      setIfDefined('web_vital.ttfb.connection_duration', attr.connectionDuration);
      setIfDefined('web_vital.ttfb.request_duration', attr.requestDuration);
      break;
    default:
      break;
  }
  return result;
};

export const createWebVitalsPlugin = (enabled: BkOTRumConfig['webVitals']): BkOTPlugin => ({
  name: 'web-vitals',
  enabled: Boolean(enabled),
  async init(context) {
    if (typeof window === 'undefined') {
      return;
    }

    const { onCLS, onFCP, onINP, onLCP, onTTFB } = await import('web-vitals/attribution');
    const clsHistogram = context.meter.createHistogram('browser.web_vital.cls', {
      unit: '1',
      description: 'Cumulative Layout Shift',
    });
    const durationHistogram = context.meter.createHistogram('browser.web_vital.duration', {
      unit: 'ms',
      description: 'Web Vitals duration metrics, including FCP, INP, LCP and TTFB',
    });

    const recordSpan = (metric: VitalLike) => {
      const span = context.startSpan('browser.web_vital', {
        ...context.config.getPageAttributes(),
        ...context.config.getMetricAttributes(),
        ...getMetricDimensions(metric),
        // 单条度量唯一 ID 仅放进 span，避免污染 metric 维度
        'web_vital.id': metric.id,
        'web_vital.value': metric.value,
        ...pickAttribution(metric, context.config.redactUrl),
      });
      span.end();
    };

    const recordDurationMetric = (metric: VitalLike) => {
      durationHistogram.record(
        metric.value,
        context.applyRedact({
          ...context.config.getPageAttributes(),
          ...context.config.getMetricAttributes(),
          ...getMetricDimensions(metric),
        })
      );
      recordSpan(metric);
    };

    onCLS(metric => {
      clsHistogram.record(
        metric.value,
        context.applyRedact({
          ...context.config.getPageAttributes(),
          ...context.config.getMetricAttributes(),
          ...getMetricDimensions(metric),
        })
      );
      recordSpan(metric);
    });
    onFCP(recordDurationMetric);
    onINP(recordDurationMetric);
    onLCP(recordDurationMetric);
    onTTFB(recordDurationMetric);
  },
});

File: bkmonitor/semconv/rum/metric/web_vitals.py
"""
Tencent is pleased to support the open source community by making 蓝鲸智云 - 监控平台 (BlueKing - Monitor) available.
Copyright (C) 2017-2025 Tencent. All rights reserved.
Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
You may obtain a copy of the License at http://opensource.org/licenses/MIT
Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
"""

from django.utils.translation import gettext_lazy as _

from semconv.constants import FieldDisplayType, FieldUnit
from semconv.rum.field import FieldSpec, RatingLevel

# ── Web Vitals 虚拟指标字段（非存储字段，由前端计算或预计算层提供）──────────────
# 注册在 SpanSpec 根级，可直接通过字段名查找，如 SpanSpec.from_field("LCP")。

#: 累积布局偏移（无单位，数值越小越好）
CLS = FieldSpec(
    field_name="CLS",
    field_alias=_("累积布局偏移"),
    rating_config=(
        RatingLevel(rating="good", value=0.1),
        RatingLevel(rating="needs_improvement", value=0.25),
        RatingLevel(rating="poor"),
    ),
)

#: 交互到下一次绘制
INP = FieldSpec(
    field_name="INP",
    field_alias=_("交互到下一次绘制"),
    field_unit=FieldUnit.MS.value,
    field_display_type=FieldDisplayType.DURATION.value,
    rating_config=(
        RatingLevel(rating="good", value=200),
        RatingLevel(rating="needs_improvement", value=500),
        RatingLevel(rating="poor"),
    ),
)

#: 最大内容绘制
LCP = FieldSpec(
    field_name="LCP",
    field_alias=_("最大内容绘制"),
    field_unit=FieldUnit.MS.value,
    field_display_type=FieldDisplayType.DURATION.value,
    rating_config=(
        RatingLevel(rating="good", value=2500),
        RatingLevel(rating="needs_improvement", value=4000),
        RatingLevel(rating="poor"),
    ),
)

#: 首次内容绘制
FCP = FieldSpec(
    field_name="FCP",
    field_alias=_("首次内容绘制"),
    field_unit=FieldUnit.MS.value,
    field_display_type=FieldDisplayType.DURATION.value,
    rating_config=(
        RatingLevel(rating="good", value=1800),
        RatingLevel(rating="needs_improvement", value=3000),
        RatingLevel(rating="poor"),
    ),
)

#: 首字节耗时
TTFB = FieldSpec(
    field_name="TTFB",
    field_alias=_("首字节耗时"),
    field_unit=FieldUnit.MS.value,
    field_display_type=FieldDisplayType.DURATION.value,
    rating_config=(
        RatingLevel(rating="good", value=800),
        RatingLevel(rating="needs_improvement", value=1800),
        RatingLevel(rating="poor"),
    ),
)
